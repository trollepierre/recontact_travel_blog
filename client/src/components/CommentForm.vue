<template>
  <form
    class="forum__form"
    @submit="submitComment">
    <p class="form-title">{{ $t("addComment") }}</p>
    <div class="input-group author">
      <label
        class="forum__label"
        for="author">
        {{ $t("name") }}
      </label>
      <input
        id="author"
        v-model="newAuthor"
        class="forum__comment-box author"
        :placeholder="anonymous">
    </div>
    <div class="input-group comment">
      <label
        class="forum__label"
        for="comment">
        {{ $t("text") }}
      </label>
      <input
        id="comment"
        v-model="newComment"
        class="forum__comment-box text"
        :placeholder="textPlaceholder">
    </div>
    <button
      type="submit"
      class="form-button">
      {{ $t("send") }}
    </button>
  </form>
</template>
<script>
  import commentsApi from '../api/comments'
  import notificationsService from '../services/notifications'

  export default {
    name: 'CommentForm',
    data() {
      return {
        newComment: '',
        newAuthor: '',
        errorComment: '',
        anonymous: this.$t('anonymous'),
        textPlaceholder: this.$t('textPlaceholder'),
        dropboxId: parseInt(this.$route.params.id, 10),
      }
    },
    methods: {
      submitComment(e) {
        e.preventDefault()
        if (this.newComment !== '') {
          const comment = {
            text: this.newComment,
            author: this.newAuthor,
          }
          return commentsApi.send(this.dropboxId, comment)
            .then(this.displaySuccessNotification)
            .catch(() => {
              this.displayErrorNotification()
            })
        }
      },
      displaySuccessNotification() {
        notificationsService.success(this, this.$t('commentSuccess'))
      },
      displayErrorNotification() {
        notificationsService.error(this, this.$t('commentError'))
      },
    },
    i18n: {
      messages: {
        fr: {
          addComment: 'Ajouter un commentaire',
          name: 'De la part de',
          commentError: 'Erreur lors de la prise en compte de ton commentaire.',
          commentSuccess: 'Ton commentaire a été pris en compte.',
          anonymous: 'Anonyme',
          textPlaceholder: 'N\'hésitez pas à ajouter des commentaires !',
          send: 'Envoyer',
          text: 'Votre message',
        },
        en: {
          addComment: 'Add a comment',
          name: 'From',
          commentError: 'Error when adding the comment.',
          commentSuccess: 'Your comment has been taken into consideration.',
          anonymous: 'Anonymous',
          textPlaceholder: 'Add your comment here!',
          send: 'Send',
          text: 'Your message',
        },
      },
    },
  }
</script>

<style>
  .forum__form {
    margin: 20px;
  }

  .form-title {
    font-size: 1.5em;
    font-weight: bold;
  }

  .input-group {
    display: flex;
    justify-content: center;
    flex-direction: column;
    margin-top: 10px;
    margin-bottom: 20px;
  }

  .forum__label {
    font-size: 1.5em;
    font-weight: bold;
    margin-bottom: 10px;
  }

  .forum__comment-box {
    font-size: 1.5em;
    align-items: start;
    border-radius: 3px;
    border: 1px solid rgba(255, 255, 255, 0.7);
    padding-left: 10px;
  }

  .forum__comment-box.text {
    height: 100px;
  }

  button.form-button {
    text-transform: uppercase;
    color: #d14800;
    background: #ffffff;
    border: 1px solid #d14800;
    cursor: pointer;
    padding: 15px 30px;
    border-radius: 4px;
    width: 100%;
    margin-bottom: 10px;
    font-weight: 700;
    max-width: 300px;
  }
</style>
