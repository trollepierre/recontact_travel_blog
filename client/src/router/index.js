import Vue from 'vue';
import Router from 'vue-router';
import ArticleList from '@/components/ArticleList';
import ArticlePage from '@/components/ArticlePage';

Vue.use(Router);

export default new Router({
  routes: [
    {
      path: '/',
      name: 'ArticleList',
      component: ArticleList,
    },
    {
      path: '/articles/:id',
      name: 'ArticlePage',
      component: ArticlePage,
    },
    {
      path: '/admin',
      name: 'ArticleList',
      component: ArticleList,
      props: { adminMode: true },
    },
  ],
});
