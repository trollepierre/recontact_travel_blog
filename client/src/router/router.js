import Vue from 'vue';
import Router from 'vue-router';
import ArticleList from '../components/ArticleList.vue';
import ArticlePage from '../components/ArticlePage.vue';
import SubscriberList from '../components/SubscriberList.vue';

Vue.use(Router);

export default new Router({
  routes: [
    {
      path: '/',
      name: 'ArticleList',
      component: ArticleList,
    },
    {
      path: '/articles/:id',
      name: 'ArticlePage',
      component: ArticlePage,
    },
    {
      path: '/admin',
      name: 'ArticleList',
      component: ArticleList,
      props: { adminMode: true },
    },
    {
      path: '/subscriptions',
      name: 'SubscriberList',
      component: SubscriberList,
    },
    { path: '/sub', redirect: '/subscriptions' },
    { path: '/a/:id', redirect: '/articles/:id' },
  ],
});
