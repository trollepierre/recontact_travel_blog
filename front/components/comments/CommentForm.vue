<template>
  <div>
    <h2 class="form__title">
      {{ $t("addComment") }}
    </h2>
    <form
      class="form"
      @submit="submitComment">
      <header class="form__header">
        <div class="input-group author">
          <label
            class="form__label author"
            for="author">
            {{ $t("name") }}
          </label>
          <input
            id="author"
            v-model="newAuthor"
            :placeholder="anonymous"
            autocomplete="off"
            spellcheck="false"
            class="forum__comment-box author">
        </div>
      </header>
      <div class="form__content">
        <div class="input-group comment">
          <label
            class="form__label comment"
            for="comment">
            {{ $t("text") }}
          </label>
          <textarea
            id="comment"
            v-model="newComment"
            :placeholder="textPlaceholder"
            class="forum__comment-box text"/>
        </div>
      </div>
      <footer class="form__footer">
        <app-button
          type="submit"
          class="form-button"
          :text="$t('send')"
          @click="submitComment"/>
      </footer>
    </form>
  </div>
</template>
<script>
  import AppButton from '@/components/AppButton/AppButton'
  import { isEmptyPlus as isEmpty } from '../../services'
  import commentsApi from '../../services/api/comments'
  import notificationsService from '../../services/services/notifications'

  export default {
    name: 'CommentForm',
    components: { AppButton },
    data() {
      return {
        newComment: '',
        newAuthor: '',
        errorComment: '',
        anonymous: this.$t('yourName'),
        textPlaceholder: this.$t('textPlaceholder'),
        dropboxId: parseInt(this.$route.params.id, 10),
      }
    },
    methods: {
      submitComment() {
        if (this.newComment !== '') {
          const comment = {
            text: this.newComment,
            author: this.newAuthor,
          }
          return commentsApi.send(this.dropboxId, comment)
            .then(this.resetNewComment)
            .then(this.reloadComments)
            .then(this.displaySuccessNotification)
            .catch(this.displayErrorNotification)
        }
        return true
      },
      reloadComments() {
        this.$emit('reload')
      },
      resetNewComment() {
        this.newComment = ''
      },
      displaySuccessNotification() {
        notificationsService.information(this.$t('commentSuccess'))
      },
      displayErrorNotification() {
        notificationsService.error(this.$t('commentError'))
      },
      isEmpty,
    },
    i18n: {
      messages: {
        fr: {
          addComment: 'Ajouter un commentaire',
          name: 'De la part de',
          yourName: 'Ton nom',
          commentError: 'Erreur lors de la prise en compte de ton commentaire.',
          commentSuccess: 'Ton commentaire a été pris en compte.',
          anonymous: 'Anonyme',
          textPlaceholder: 'N\'hésitez pas à ajouter des commentaires !',
          send: 'Envoyer',
          text: 'Votre message',
        },
        en: {
          addComment: 'Add a comment',
          name: 'From',
          yourName: 'Your name',
          commentError: 'Error when adding the comment.',
          commentSuccess: 'Your comment has been taken into consideration.',
          anonymous: 'Anonymous',
          textPlaceholder: 'Add your comment here!',
          send: 'Send',
          text: 'Your message',
        },
      },
    },
  }
</script>

<style lang="scss" scoped>
  .form__title {
    font-size: 1.5em;
    font-weight: bold;
    font-family: serif;
  }

  .form {
    min-width: 260px;
    background: $comment-form-bg;
    border-radius: 4px;
    border: 1px solid $border;
    display: flex;
    flex-direction: column;
    width: 50%;
    margin-left: auto;
    margin-right: auto;
    margin-bottom: 30px;
  }

  .form__header, .form__content {
    padding: 15px;
    height: auto;
    display: flex;
    justify-content: center;
    align-items: center;
  }

  .forum__comment-box {
    font-family: serif;
    font-size: 16px;
    font-weight: 400;
    margin: 0;
    overflow-wrap: break-word;
    padding-left: 10px;
    border: 1px solid $border;
    background-color: $comment-form-input-bg;
    border-radius: 4px;
    color: $comment-form-color;
  }

  .input-group {
    display: flex;
    justify-content: center;
    flex-direction: column;
    margin-bottom: 5px;
    width: 80%;
  }

  .input-group.author {
    width: 60%;
    max-width: 300px;
  }

  .form__label {
    font-size: 16px;
    font-weight: 700;
    font-family: serif;
    color: $comment-form-label;
    margin: 0;
    overflow-wrap: break-word;
    padding-bottom: 10px;
  }

  .forum__comment-box.text {
    font-weight: 400;
    height: 100px;
    word-break: keep-all;
    padding: 10px;
  }

  .form__footer {
    text-align: center;
    padding: 15px;
    border-top: 1px solid $border;
  }

  .form-button {
    max-width: 300px;
  }

  @media only screen and (max-width: 1000px) {
    .form {
      width: 100%;
    }
  }
</style>
