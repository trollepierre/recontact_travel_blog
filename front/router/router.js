import Router from 'vue-router'
import Vue from 'vue'
import ArticlePage from '../components/ArticlePage/ArticlePage.vue'

Vue.use(Router)

export const conf = {
  mode: 'history',
  routes: [
    // {
    //   path: '/',
    //   name: 'ArticleList',
    //   component: ArticleList,
    // },
    {
      path: '/articles/:id',
      name: 'ArticlePage',
      component: ArticlePage,
    },
    // {
    //   path: '/admin',
    //   name: 'ArticleList',
    //   component: ArticleList,
    //   props: { adminMode: true },
    // },
    // {
    //   path: '/subscriptions',
    //   name: 'SubscriberList',
    //   component: SubscriberList,
    // },
    { path: '/sub', redirect: '/subscriptions' },
    { path: '/a/:id', redirect: '/articles/:id' },
  ],
}

export default new Router(conf)

