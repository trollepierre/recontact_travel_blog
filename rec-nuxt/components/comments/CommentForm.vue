<template>
  <div class="comment__form">
    <p class="form__title">
      {{ $t("addComment") }}
    </p>
    <form
      class="form"
      @submit="submitComment">
      <header class="form__header">
        <div class="input-group author">
          <label
            v-if="!isEmpty(newAuthor)"
            class="form__label author"
            for="author">
            {{ $t("name") }}
          </label>
          <input
            id="author"
            v-model="newAuthor"
            :placeholder="anonymous"
            class="forum__comment-box author">
        </div>
      </header>
      <main class="form__content">
        <div class="input-group comment">
          <label
            class="form__label comment"
            for="comment">
            {{ $t("text") }}
          </label>
          <textarea
            id="comment"
            v-model="newComment"
            :placeholder="textPlaceholder"
            class="forum__comment-box text"/>
        </div>
      </main>
      <footer class="form__footer">
        <button
          type="submit"
          class="form-button">
          {{ $t("send") }}
        </button>
      </footer>
    </form>
  </div>
</template>
<script>
  import { isEmpty } from 'ramda'
  import commentsApi from '../../services/api/comments'
  import notificationsService from '../../services/services/notifications'

  export default {
    name: 'CommentForm',
    data() {
      return {
        newComment: '',
        newAuthor: '',
        errorComment: '',
        anonymous: this.$t('name'),
        textPlaceholder: this.$t('textPlaceholder'),
        dropboxId: parseInt(this.$route.params.id, 10),
      }
    },
    methods: {
      submitComment(e) {
        e.preventDefault()
        if (this.newComment !== '') {
          const comment = {
            text: this.newComment,
            author: this.newAuthor,
          }
          return commentsApi.send(this.dropboxId, comment)
            .then(this.resetNewComment)
            .then(this.reloadComments)
            .then(this.displaySuccessNotification)
            .catch(this.displayErrorNotification)
        }
      },
      reloadComments() {
        this.$emit('reload')
      },
      resetNewComment() {
        this.newComment = ''
      },
      displaySuccessNotification() {
        notificationsService.success(this, this.$t('commentSuccess'))
      },
      displayErrorNotification() {
        notificationsService.error(this, this.$t('commentError'))
      },
      isEmpty,
    },
    i18n: {
      messages: {
        fr: {
          addComment: 'Ajouter un commentaire',
          name: 'De la part de',
          commentError: 'Erreur lors de la prise en compte de ton commentaire.',
          commentSuccess: 'Ton commentaire a été pris en compte.',
          anonymous: 'Anonyme',
          textPlaceholder: 'N\'hésitez pas à ajouter des commentaires  !',
          send: 'Envoyer',
          text: 'Votre message',
        },
        en: {
          addComment: 'Add a comment',
          name: 'From',
          commentError: 'Error when adding the comment.',
          commentSuccess: 'Your comment has been taken into consideration.',
          anonymous: 'Anonymous',
          textPlaceholder: 'Add your comment here!',
          send: 'Send',
          text: 'Your message',
        },
      },
    },
  }
</script>

<style scoped>
  .form__title {
    font-size: 1.5em;
    font-weight: bold;
  }

  .form {
    min-width: 260px;
    background: #ffffff;
    border-radius: 4px !important;
    box-shadow: 0 1px 1px rgba(0, 0, 0, .15);
    border: 1px solid rgba(0, 0, 0, .09);
    display: flex;
    flex-direction: column;
    color: #535a60;
    width: 50%;
    margin-left: auto;
    margin-right: auto;
    margin-bottom: 30px;
  }

  .form__header, .form__content {
    border-bottom: 1px solid #e6e6e6;
    padding: 15px;
    height: auto;
    display: flex;
    justify-content: center;
    align-items: center;
  }

  .forum__comment-box {
    font-size: 16px;
    font-weight: 400;
    margin: 0;
    overflow-wrap: break-word;
    padding-left: 10px;
    border-width: 1px;
    border-style: solid;
    border-color: gray;
    border-image: initial;
  }

  .input-group {
    display: flex;
    justify-content: center;
    flex-direction: column;
    margin-bottom: 5px;
    width: 80%;
  }

  .input-group.author {
    width: 60%;
    max-width: 300px;
  }

  .form__label {
    font-size: 16px;
    font-weight: 700;
    color: #07c;
    margin: 0;
    overflow-wrap: break-word;
    padding-bottom: 10px;
  }

  .forum__comment-box.text {
    font-weight: 400;
    height: 100px;
    word-break: keep-all;
    padding: 10px;
  }

  .form__footer {
    text-align: center;
    padding: 15px;
    border-top: 1px solid #e6e6e6;
  }

  button.form-button {
    text-transform: uppercase;
    color: #d14800;
    background: #ffffff;
    border: 1px solid #d14800;
    cursor: pointer;
    padding: 15px 30px;
    border-radius: 4px;
    width: 100%;
    margin-bottom: 10px;
    font-weight: 700;
    max-width: 300px;
  }

  @media only screen and (max-width: 1000px) {
    .form {
      width: 100%;
    }
  }
</style>
