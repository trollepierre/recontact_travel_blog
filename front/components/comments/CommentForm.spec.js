import VueRouter from 'vue-router' // eslint-disable-line import/no-extraneous-dependencies
import VueI18n from 'vue-i18n'

import CommentForm from './CommentForm.vue'
import router from '../../test/router/router'
import commentsApi from '../../services/api/comments'
import notificationsService from '../../services/services/notifications'

describe('Component | CommentForm.vue', () => {
  let localVue
  let wrapper

  const dropboxId = '8'
  const commentsFromApi = [{ text: 'comment1' }]

  beforeEach(() => {
    commentsApi.fetch = jest.fn()
    commentsApi.fetch.mockResolvedValue(commentsFromApi)
    console.warn = jest.fn()

    localVue = createLocalVue()
    localVue.use(VueI18n)
    localVue.use(VueRouter)

    wrapper = shallowMount(CommentForm, { localVue, router, data: () => ({ dropboxId }) })
  })

  it('should be named "CommentForm"', () => {
    expect(wrapper.name()).toEqual('CommentForm')
  })

  describe('template', () => {
    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('methods', () => {
    describe('submitComment', () => {
      const newComment = 'my text'
      beforeEach(() => {
        commentsApi.send = jest.fn()
        notificationsService.information = jest.fn()
        notificationsService.error = jest.fn()
      })

      it('should send comment to api', () => {
        // Given
        commentsApi.send.mockResolvedValue({ text: 'createdComment' })
        wrapper = shallowMount(CommentForm, { localVue, router, data: () => ({ dropboxId, newComment }) })

        // When
        wrapper.vm.submitComment({ preventDefault: jest.fn() })

        // Then
        expect(commentsApi.send).toHaveBeenCalledOnceWith(dropboxId, { author: '', text: newComment })
      })

      it('should display success notification', async () => {
        // Given
        commentsApi.send.mockResolvedValue({ text: 'createdComment' })
        wrapper = shallowMount(CommentForm, { localVue, router, data: () => ({ dropboxId, newComment }) })

        // When
        await wrapper.vm.submitComment({ preventDefault: jest.fn() })

        // Then
        expect(notificationsService.information).toHaveBeenCalledOnceWith('Your comment has been taken into consideration.')
      })

      describe('when api throws error', () => {
        it('should display error notification', async () => {
          // Given
          commentsApi.send.mockRejectedValue({ error: 'no comment...' })
          wrapper = shallowMount(CommentForm, { localVue, router, data: () => ({ dropboxId, newComment }) })

          // When
          await wrapper.vm.submitComment({ preventDefault: jest.fn() })

          // Then
          expect(notificationsService.error).toHaveBeenCalledOnceWith('Error when adding the comment.')
        })
      })

      describe('when newComment is not set', () => {
        it('should not send comment', () => {
          // Given
          wrapper = shallowMount(CommentForm, { localVue, router, data: () => ({ dropboxId, newComment: '' }) })

          // When
          wrapper.vm.submitComment({ preventDefault: jest.fn() })

          // Then
          expect(commentsApi.send).not.toHaveBeenCalled()
        })
      })
    })
  })

  describe('locales', () => {
    const languages = Object.keys(CommentForm.i18n.messages)

    it('contains 2 languages', () => {
      expect(languages).toHaveLength(2)
      expect(languages).toEqual(['fr', 'en'])
    })

    describe('each language', () => {
      describe('fr', () => {
        const locales = Object.keys(CommentForm.i18n.messages.fr)

        it('contains 8 locales', () => {
          expect(locales).toHaveLength(8)
          expect(locales).toEqual([
            'addComment',
            'name',
            'commentError',
            'commentSuccess',
            'anonymous',
            'textPlaceholder',
            'send',
            'text',
          ])
        })
      })

      describe('en', () => {
        const locales = Object.keys(CommentForm.i18n.messages.en)

        it('contains 8 locales', () => {
          expect(locales).toHaveLength(8)
          expect(locales).toEqual([
            'addComment',
            'name',
            'commentError',
            'commentSuccess',
            'anonymous',
            'textPlaceholder',
            'send',
            'text',
          ])
        })
      })
    })
  })
})
