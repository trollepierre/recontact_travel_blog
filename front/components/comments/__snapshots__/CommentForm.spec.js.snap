// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component | CommentForm.vue template should match snapshot 1`] = `
<div
  class="comment__form"
>
  <p
    class="form__title"
  >
    
    addComment
  
  </p>
   
  <form
    class="form"
  >
    <header
      class="form__header"
    >
      <div
        class="input-group author"
      >
        <!---->
         
        <input
          class="forum__comment-box author"
          id="author"
          placeholder="name"
        />
      </div>
    </header>
     
    <main
      class="form__content"
    >
      <div
        class="input-group comment"
      >
        <label
          class="form__label comment"
          for="comment"
        >
          
          text
        
        </label>
         
        <textarea
          class="forum__comment-box text"
          id="comment"
          placeholder="textPlaceholder"
        />
      </div>
    </main>
     
    <footer
      class="form__footer"
    >
      <button
        class="form-button"
        type="submit"
      >
        
        send
      
      </button>
    </footer>
  </form>
</div>
`;
